Feature/bt 165 invulnerables

roy-powerfinance opened the pull request:
Added loggings to when:
* invulnerable validator is NOT slashed
* invulnerable validators are modified

Added a test that verified that  invulnerable validators can be changed via a function call
Added a test that verified that invulnerable validators can only be changed with root permission

roy-powerfinance committed:
Added logging when an invulnerable validator was not slashed
Added logging when invulnerables are changed

roy-powerfinance committed:
Added a test to check that the invulnerable set can be configured via
function calls

roy-powerfinance committed:
Added a test that ensures that only root accounts can change invulnerable
validator sets

roy-powerfinance committed:
Merge remote-tracking branch 'origin/develop' into feature/bt_165_invulnerables

* origin/develop:
  Fix staking reward functions to use T::RewardCurrency (#161)

roy-powerfinance committed:
fixed a build error

roy-powerfinance committed:
fixed formatting

cowboy-bebug commented on crml/staking/src/lib.rs:
```diff
@@ -1191,7 +1191,9 @@ decl_module! {
 			ensure_root(origin)?;
-			<Invulnerables<T>>::put(validators);
+			<Invulnerables<T>>::put(validators.clone());
+			validators.iter().for_each( |v| debug::print!("Set invulnerable:{:?}", v) );
```
[Option] Maybe something like this is also okay (just to avoid printing many lines): `debug::print!("Set invulnerables: {:?}", validators);`?

roy-powerfinance replied:
yeah true

hoani replied:
I'm assuming that each invulnerable is a 32-byte address? the print may look messy no matter how we print it. It would be awesome if we could do something like "Set invulnerables: 4ab8896...4c, 6ccb23...67", but that's definately out of scope =D

roy-powerfinance replied:
changed this to a single call

cowboy-bebug commented on crml/staking/src/tests.rs:
```diff
@@ -2313,6 +2313,68 @@ fn invulnerables_are_not_slashed() {
+
+		//Changing the invulnerables with root access.
+		let _ = Staking::set_invulnerables(Origin::ROOT, vec![21, 11]);
+		assert_eq!(Staking::invulnerables(), vec![21, 11]);
```
Can we add something like below to show that setting it empty is also possible:
```rust
let _ = Staking::set_invulnerables(Origin::ROOT, vec![]);
assert_eq!(Staking::invulnerables(), vec![]); // TODO: may want to control this logic
```

roy-powerfinance replied:
Yep added a new test for this

cowboy-bebug reviewed (commented):
LGTM! Could u also add an event for `set_invulerables`?

roy-powerfinance committed:
Added a test that verifies the invulnerables can be set as empty
Made some improvements as suggested by the PR review
Added 2 new events - when invulnerables are set and when invulnerables
are not slashed

cowboy-bebug commented on crml/staking/src/lib.rs:
```diff
@@ -1191,7 +1196,10 @@ decl_module! {
-			<Invulnerables<T>>::put(validators);
+			<Invulnerables<T>>::put(validators.clone());
+			debug::print!("Set invulnerable:{:?}", validators.clone() );
+			Self::deposit_event(RawEvent::SetInvulnerables(validators.clone()) );
```
[Optional] last clone can be removed

jordy25519 replied:
`debug::print!("Set invulnerable:{:?}", validators.clone() );` does this need clone?

roy-powerfinance replied:
Yep removed both clone()

cowboy-bebug commented on crml/staking/src/tests.rs:
```diff
@@ -2313,6 +2313,82 @@ fn invulnerables_are_not_slashed() {
 }
 
+#[test]
+fn invulnerables_can_be_set() {
```
[Optional] Do we really need to add this test since it's very similar to `fn invulnerables_are_not_slashed` and `fn invulnerables_can_only_be_set_by_root` already covers it

roy-powerfinance replied:
the other 2 new tests only check if storage is correct, but do not check for invulnerability.
this is sort of a "end to end" test that tests the whole thing. 
I don't mind removing it though

cowboy-bebug reviewed (approved):
LGTM

jordy25519 reviewed (commented):
lgtm!

roy-powerfinance committed:
removed some redundant clone() calls
